### PR TITLE
Ecl sum keyword vector

### PR DESCRIPTION
**Pull request by joakim-hove**
**Task**
When exporting multiple cases in tabular form it is essential that they have the same keys. With this PR we will emit an empty string for missing keys, to ensure that alignment is correct.

Part of: ERT-1435

**Pre un-WIP checklist**
- [x] Statoil tests pass locally

### CODE REVIEW

**Comment by berland** [comment]
Is this commit also handling the case when the number of parameters from parameters.txt vary between realizations?

**Comment by joakim-hove** [comment]
> Is this commit also handling the case when the number of parameters from parameters.txt vary between realizations?

No - this code is not related to `parameters.txt` at all - but the feature request is noted.

File: lib/ecl/ecl_sum_data.c [hunk 1]
```diff
@@ -1205,25 +1203,40 @@ void ecl_sum_data_fwrite_interp_csv_line(const ecl_sum_data_type * data, time_t
...
   for (int i = 0; i < num_keywords; i++) {
-    double value = ecl_sum_data_vector_iget( data, sim_time, keylist , i , time_index1, time_index2, weight1, weight2);
-    double_vector_iset( results, i , value );
+    int params_index = ecl_sum_vector_iget_param_index(keylist, i);
+    if (params_index >= 0) {
+      bool is_rate = ecl_sum_vector_iget_is_rate(keylist, i);
+      double value = ecl_sum_data_vector_iget( data, sim_time, params_index, is_rate, time_index1, time_index2, weight1, weight2);
+      double_vector_iset( results, i , value );
+    } else
+      double_vector_iset( results, i , missing_value);
```
Comment (pgdr) [inline lib/ecl/ecl_sum_data.c]: I'm not sure I understand this part (I'm sure I don't understand this part) but is it safe to set the values to -1?  Is it necessary?

Comment (joakim-hove) [reply lib/ecl/ecl_sum_data.c]: > I'm not sure I understand this part (I'm sure I don't understand this part) but is it safe to set the values to -1? Is it necessary?

Good question - I was slightly uncomfortable doing that. The point of the whole PR is that if the keyword vector asks for keywords: `["KEY1", "KEY2", "KEY3"]` and the summary instance in question is missing `KEY2` we still need to return a result with three elements - to ensure that things are correctly aligned. As matters stand it is actyally the `_fprintf( )` function which is the most important - in that case the printed result should:

```txt
100,,200
```

i.e. an empty element in the csv output. We could:

1. [Maybe] just remove the offending function.
2. Pass in a user supplied value to be used in the case of missing values.
3. Use a different value than `-1` - e.g. `nan`.
4. Return an additional mask-like structure to indicate which values are valid.

File: lib/ecl/ecl_sum_vector.c [hunk 2]
```diff
@@ -58,6 +58,27 @@ ecl_sum_vector_type * ecl_sum_vector_alloc(const ecl_sum_type * ecl_sum){
...
+  for (int i=0; i < stringlist_get_size(vector->key_list); i++) {
+    const char * key = stringlist_iget(vector->key_list, i);
+    if (ecl_sum_has_general_var(ecl_sum, key))
+      ecl_sum_vector_add_key(new_vector, key);
+    else
+      ecl_sum_vector_add_invalid_key(new_vector, key);
+  }
+  return new_vector;
+}
+
```
Comment (pgdr) [inline lib/ecl/ecl_sum_vector.c]: Do I understand correctly if `params_index < 0` then the element is invalid?

If so, should maybe have an `ecl_sum_vector_is_valid(const ecl_sum_vector_type * vec, int i)` that returns true if and only if `vec->node_index[i] < 0`?

File: lib/ecl/ecl_sum_vector.c [hunk 3]
```diff
@@ -58,6 +58,27 @@ ecl_sum_vector_type * ecl_sum_vector_alloc(const ecl_sum_type * ecl_sum){
...
+static void ecl_sum_vector_add_invalid_key(ecl_sum_vector_type * vector, const char * key) {
+  int_vector_append(vector->node_index_list, -1);
+  bool_vector_append(vector->is_rate_list, false);
+  stringlist_append_copy(vector->key_list, key);
+}
+
+
+
+ecl_sum_vector_type * ecl_sum_vector_alloc_layout_copy(const ecl_sum_vector_type * vector, const ecl_sum_type * ecl_sum) {
+  ecl_sum_vector_type * new_vector = ecl_sum_vector_alloc(ecl_sum);
```
Comment (pgdr) [inline lib/ecl/ecl_sum_vector.c]: Shouldn't this use `vector->ecl_sum` and not take `ecl_sum` as an argument?  Or perhaps I don't understand this function.   Can you add a comment?

File: python/tests/ecl/test_sum.py [hunk 4]
```diff
@@ -287,23 +288,32 @@ def test_invalid(self):
...
-        t = case.getDataStartTime( ) + datetime.timedelta( days = 43 );
-        data = case.get_interp_row( kw_list , t )
-        for d1,d2 in zip(data, [ case.get_interp("FOPT", date = t),
-                                 case.get_interp("FOPT", date = t),
-                                 case.get_interp("FOPT", date = t) ]):
+        t = case1.getDataStartTime( ) + datetime.timedelta( days = 43 );
+        data = case1.get_interp_row( kw_list , t )
+        for d1,d2 in zip(data, [ case1.get_interp("FOPT", date = t),
+                                 case1.get_interp("FOPT", date = t),
+                                 case1.get_interp("FOPT", date = t) ]):
```
Comment (pgdr) [inline python/tests/ecl/test_sum.py]: I have a question regarding `get_interp_row`, but it doesn't really belong to this PR.  But that function in `c` seems to link `ecl_sum` back to `ecl_sum_vector`.  Could it not be placed in `ecl_sum_vector` instead?

File: lib/ecl/ecl_sum_vector.c (hunk 2 above)
Comment (joakim-hove) [reply lib/ecl/ecl_sum_vector.c]: > Do I understand correctly if params_index < 0 then the element is invalid?

Yes

> should maybe have an ecl_sum_vector_is_valid(const ecl_sum_vector_type * vec, int i)

YAGNI??

File: lib/ecl/ecl_sum_vector.c (hunk 3 above)
Comment (joakim-hove) [reply lib/ecl/ecl_sum_vector.c]: > Can you add a comment?

Yes - I will add a comment.

File: lib/ecl/ecl_sum_data.c [hunk 5]
```diff
@@ -1205,25 +1203,40 @@ void ecl_sum_data_fwrite_interp_csv_line(const ecl_sum_data_type * data, time_t
   ecl_sum_data_init_interp_from_sim_time(data, sim_time, &time_index1, &time_index2, &weight1, &weight2);
 
   for (int i = 0; i < num_keywords; i++) {
-    double value = ecl_sum_data_vector_iget( data, sim_time, keylist , i , time_index1, time_index2, weight1, weight2);
+    int params_index = ecl_sum_vector_iget_param_index(keylist, i);
+    if (params_index >= 0) {
```
Comment (pgdr) [inline lib/ecl/ecl_sum_data.c]: I need it here?  We don't need to _expose_ the function out of this file, but I am confused as to whether `params_index=0` is valid or invalid.  If we had an "iget valid" function, that would be beyond doubt?

File: python/tests/ecl/test_sum.py (hunk 4 above)
Comment (joakim-hove) [reply python/tests/ecl/test_sum.py]: I have not looked at this code right now - but I *think* you are mxing two different types of vectors:

`ecl_sum_vector`: is one keyword, all timesteps.
`ecl_sum_keyword_vector`: A vector/collection of keywords, which are typically used to evaluate many keywords at the sime time point [which has a significant startup cost].

File: lib/ecl/ecl_sum_data.c (hunk 5 above)
Comment (joakim-hove) [reply lib/ecl/ecl_sum_data.c]: >  If we had an "iget valid" function, that would be beyond doubt?

Sure - I will add it.

**Comment by joakim-hove** [comment]
OK - fixup with:
1. Added function to check if index is valid.
2. Comments 
3. Removed setting of invalid value - user should set that beforehand.
